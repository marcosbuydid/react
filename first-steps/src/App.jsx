import { useState, Fragment } from "react";
import PropTypes from 'prop-types';

export const App = ({ title, counter }) => {

    const [value, setValue] = useState(counter);

    const increment = () => {
        setValue(value + 1);
    }

    const decrement = () => {
        setValue(value - 1);
    }

    const reset = () => {
        setValue(0);
    }

    return (
        /*
        Fragments are used to group multiple elements together. 
        You can put multiple elements in any place where 
        a single element can go.
        */
        <Fragment>
            <h1>{title}</h1>
            <p>{value}</p>
            <button onClick={increment}>
                Increment
            </button>
            <button onClick={decrement}>
                Decrement
            </button>
            <button onClick={reset}>
                Reset
            </button>
        </Fragment >
    );
}

//here we declare default props
App.defaultProps = {
    title: 'CounterApp',
    counter: 0,
}

//here we declare what type are the props we define
App.propTypes = {
    title: PropTypes.string.isRequired,
    counter: PropTypes.number.isRequired,
}