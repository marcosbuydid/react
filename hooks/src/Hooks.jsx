import React from 'react'

export const Hooks = () => {
    return (
        <h1>Hooks</h1>
    )
}
