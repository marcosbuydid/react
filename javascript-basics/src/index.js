import { restaurants } from './data/restaurants';

//Constants
const name = 'Your name';
const lastName = 'Your lastName';

//Variables
let age = 34;
age = 35;

//Function
export function getFullInformation(name, lastName, age) {
    return name + ' ' + lastName + ' ' + age;
}

const getPersonDetails = (name, lastName, age) => {
    return name + ' ' + lastName + ' ' + age;
}

console.log((getFullInformation(name, lastName, age)));
console.log((getPersonDetails('Lara', 'Bisnett', 27)));

//literal objects
const person = {
    name: 'Tim',
    lastName: 'Szas',
    age: 29,
    address: {
        street: 'Rockaway Blvd/124 St',
        zip: 11420,
        state: 'New York',
        country: 'United States',
    }
}

//clone object
const personTwo = { ...person };
personTwo.name = 'Cedrick';

console.log(person);
console.log(personTwo);

//array
const array = [1, 2, 3, 4];

const array2 = [...array, 5];

console.log(array2);

//object destructure

//rename variables because are already in use
const { name: name2, age: age2, address } = person;
console.log(name2);
console.log(age2);
console.log(address);

export const getAddressInformation = (person) => {
    return person.address;
}

//getInformation(person);

const personNames = ['Nick', 'Michael', 'Peter'];
console.log(personNames[1]);

const useState = (value) => {
    return [value, () => { console.log('Hello World') }];
}

//import export and common array functions
const getRestaurantById = (id) => {
    return (restaurants).find((restaurant) => restaurant.id === id);
}

const filterRestaurantByOwner = (owner) => {
    return (restaurants).filter((restaurant) => restaurant.owner.includes(owner));
}

console.log(getRestaurantById(4));

console.log(filterRestaurantByOwner('Dominique'));

//Promises
const getRestaurantByIdAsync = (id) => {

    return new Promise((resolve, reject) => {
        setTimeout(() => {
            const restaurant = getRestaurantById(id);
            if (restaurant) {
                resolve(restaurant);
            }
            else {
                reject('Cannot find the restaurant');
            }
        }, 2000)
    });

};

getRestaurantByIdAsync(5)
    .then((restaurant) => console.log('restaurant', restaurant))
    .catch(error => console.warn(error));


//fetch API using giphy.com
const apiKey = 'OTqQ2Jl2JvdRzZtZylYRmuPeGjSTCo4z';
const petition = fetch(`https://api.giphy.com/v1/gifs/random?api_key=${apiKey}`);

petition
    .then(response => response.json())
    .then(({ data }) => {
        const { url } = data.images.original;

        const img = document.createElement('img');
        img.src = url;
        document.body.append(img);

    })
    .catch(console.warn)


//async and await

/*
When defining a function as async we can use the keyword "await"
before of any expression that returns a promise. By doing that
the external function (async) will pause until the promise is resolved.
The keyword await receives a promise and convert it to a return value
(or generate an exception error).
When using await, javascypt will wait until the promise finish. If is
completed with success the obtained value is returned, else if is 
rejected, an error is throwed by the exception.
*/

const getImage = async () => {
    try {
        const apiKey = 'OTqQ2Jl2JvdRzZtZylYRmuPeGjSTCo4z';
        const response = await fetch(`https://api.giphy.com/v1/gifs/random?api_key=${apiKey}`);
        const { data } = await response.json();
        const { url } = data.images.original;
        const img = document.createElement('img');
        img.src = url;
        document.body.append(img);
    } catch (error) {
        console.error(error);
    }
}

getImage();
