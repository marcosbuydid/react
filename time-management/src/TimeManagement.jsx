export const TimeManagement = () => {
    return (
        <h3>Time Management</h3>
    )
}
