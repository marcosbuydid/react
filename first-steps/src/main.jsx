import React from 'react';
import ReactDOM from 'react-dom/client';
import { App } from './App';
import './styles.css';

ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        <App propOne="This is propOne" propTwo="This is propTwo" />
    </React.StrictMode>
);