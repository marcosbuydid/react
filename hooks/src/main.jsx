import React from 'react';
import ReactDOM from 'react-dom/client';
import { Hooks } from './Hooks';
import { CounterApp } from './useState/CounterApp'
import { CounterAppCustom } from './useState/CounterAppCustom'
import { SimpleForm } from './useEffect/SimpleForm'
import { CustomForm } from './useEffect/CustomForm'
import './index.css';
import { FocusScreen } from './useRef/FocusScreen';
import { Memorize } from './memo/Memorize';
import { MemorizeCustom } from './memo/MemorizeCustom';
import { CallBack } from './memo/CallBack';
import { TaskManager } from './useReducer/TaskManager';

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    {/* <Hooks /> */}
    {/* <CounterApp /> */}
    {/* <CounterAppCustom /> */}
    {/* <SimpleForm /> */}
    {/* <CustomForm /> */}
    {/* <FocusScreen /> */}
    {/* <Memorize /> */}
    {/* <MemorizeCustom /> */}
    {/* <CallBack /> */}
    <TaskManager />


  </React.StrictMode>
)
