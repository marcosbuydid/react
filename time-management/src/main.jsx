import React from 'react'
import ReactDOM from 'react-dom/client'
import './styles.css'
import { TimeManagement } from './TimeManagement.jsx'

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    <TimeManagement />
  </React.StrictMode>,
)
