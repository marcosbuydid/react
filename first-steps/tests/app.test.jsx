/** @jest-environment jsdom */

import { fireEvent, render, screen } from '@testing-library/react';
import { App } from '../src/App';
import React from 'react';

describe('Tests on <App />', () => {

    const title = 'CounterApp';
    const initialValue = 0;

    test('container must match with the snapshot', () => {
        const { container } = render(<App title={title} counter={initialValue} />)
        expect(container).toMatchSnapshot();
    })

    test('initial value of 100 must be showed', () => {
        render(<App title={title} counter={100} />)
        expect(screen.getByText('100'));
    })

    test('initial value must increase when increment button is pressed', () => {
        render(<App title={title} counter={initialValue} />)
        fireEvent.click(screen.getByText('Increment'))
        expect(screen.getByText('1')).toBeTruthy();
    })

    test('initial value must decrease when decrement button is pressed', () => {
        render(<App title={title} counter={initialValue} />)
        fireEvent.click(screen.getByText('Decrement'))
        expect(screen.getByText('-1')).toBeTruthy();
    })

    test('initial value must be 0 when reset button is pressed', () => {
        render(<App title={title} counter={120} />)
        fireEvent.click(screen.getByText('Reset'))
        expect(screen.getByText('0')).toBeTruthy();
    })
});