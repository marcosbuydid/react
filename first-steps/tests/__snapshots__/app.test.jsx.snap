// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <App /> container must match with the snapshot 1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <p>
    0
  </p>
  <button>
    Increment
  </button>
  <button>
    Decrement
  </button>
  <button>
    Reset
  </button>
</div>
`;
